Source/EventAnalysis/KTIterativeTrackClustering.cc

```diff
@@ -0,0 +1,374 @@
+    {
+        if (! FindMatchingTracks())
+        {
+            KTERROR(itclog, "An error occurred while identifying extrapolated tracks");
+            return false;
+        }
+
+        KTDEBUG(itclog, "Track building complete");
+        fDoneSignal();
+
+        return true;
+    }
+
+    bool KTIterativeTrackClustering::FindMatchingTracks()
+    {
+        KTINFO(itclog, "Finding extrapolated tracks");
+        KTDEBUG(itclog, "TimeGapTolerance FrequencyAcceptance and MaxTrackWidth are: "<<fTimeGapTolerance<< " "<<fFrequencyAcceptance<< " "<<fMaxTrackWidth);
+        fNewTracks.clear();
+
+        unsigned NumberOfTracks = fCompTracks.size();
+        unsigned NumberOfNewTracks = fNewTracks.size();
+
+
+        if (NumberOfTracks > 1)
+        {
+            while (NumberOfTracks!=NumberOfNewTracks)
+            {
+                NumberOfTracks = fCompTracks.size();
+                KTDEBUG(itclog, "Number of tracks to cluster: "<< NumberOfTracks);
+                this->ExtrapolateClustering();
+
+                // Update number of tracks
+                NumberOfNewTracks = fNewTracks.size();
+
+                KTDEBUG(itclog, "Number of new tracks: "<< NumberOfNewTracks);
+
+                fCompTracks.clear();
+                fCompTracks = fNewTracks;
+                fNewTracks.clear();
+            }
+        }
+
+        this->EmitTrackCandidates();
+
+        return true;
+    }
+
+
+    bool KTIterativeTrackClustering::ExtrapolateClustering()
+    {
+        bool match;
+        for (std::vector<KTProcessedTrackData>::iterator compIt = fCompTracks.begin(); compIt != fCompTracks.end(); ++compIt)
+        {
+            match = false;
+            for (std::vector<KTProcessedTrackData>::iterator newIt = fNewTracks.begin(); newIt != fNewTracks.end(); ++newIt)
+            {
+                if (this->DoTheyMatch(*compIt, *newIt))
+                {
+                    match = true;
+                    KTDEBUG(itclog, "Found matching tracks");
+                    this->CombineTracks(*compIt, *newIt);
+                    break;
+                }
+                // it is possible that the segments that get combined first are not direct neighbors in time
+                // in that case there can be a track segment very close to an already combined track
+                if (this->DoTheyOverlap(*compIt, *newIt))
+                {
+                    match = true;
+                    KTDEBUG(itclog, "Found overlapping tracks");
+                    this->CombineTracks(*compIt, *newIt);
+                    break;   
+                }
+            }
+
+            if (match == false)
+            {
+                KTProcessedTrackData newTrack(*compIt);
+                fNewTracks.push_back(newTrack);
+            }   
+        }
+        return true;
+    }
+
+
+    void KTIterativeTrackClustering::CombineTracks(KTProcessedTrackData& oldTrack, KTProcessedTrackData& newTrack)
```

Comment:
`oldTrack` can be a const reference, since nothing in it is set.

Also, this function can be `const`, i believe